examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py

```diff
@@ -0,0 +1,590 @@
+        )
+
+    def call(self, inputs):
+        original_output = self.original_layer(inputs)
+        if self.trainable:
+            # If we are fine-tuning the model, we will add LoRA layers' output
+            # to the original layer's output.
+            lora_output = self.B(self.A(inputs)) * self._scale
+            return original_output + lora_output
+
+        # If we are in inference mode, we "merge" the LoRA layers' weights into
+        # the original layer's weights - more on this in the text generation
+        # section!
+        return original_output
+
+
+"""
+### Inject LoRA layer into the model
+
+We will now hack the original GPT-2 model and inject LoRA layers into it. Let's
+do a couple of things before doing that:
+
+- Delete previous model;
+- Reset "peak" GPU memory usage using `tf.config.experimental.reset_memory_stats`;
+- Load a new GPT-2 model.
+"""
+
+del gpt2_lm
+del optimizer
+del loss
+del learning_rate
+
+# This resets "peak" memory usage to "current" memory usage.
+tf.config.experimental.reset_memory_stats("GPU:0")
+# We will subtract this from the total memory usage when calculating the
+# memory usage for LoRA.
+current_memory_usage = tf.config.experimental.get_memory_info("GPU:0")
+
+# Load the original model.
+preprocessor = keras_nlp.models.GPT2CausalLMPreprocessor.from_preset(
+    "gpt2_base_en",
+    sequence_length=128,
+)
+lora_model = keras_nlp.models.GPT2CausalLM.from_preset(
+    "gpt2_base_en",
+    preprocessor=preprocessor,
+)
+
+"""
+We will now override the original query/value projection matrices with our
+new LoRA layers.
+"""
+
+gpt2_backbone_layer_name = [
+    layer.name for layer in list(lora_model.layers) if "GPT2Backbone" in str(layer)
+][0]
+
+for layer_idx in range(lora_model.get_layer(gpt2_backbone_layer_name).num_layers):
+    # Change query dense layer.
+    original_query_layer = (
+        lora_model.get_layer(gpt2_backbone_layer_name)
+        .get_layer(f"transformer_layer_{layer_idx}")
+        ._self_attention_layer._query_dense
+    )
+
+    (
+        lora_model.get_layer(gpt2_backbone_layer_name)
+        .get_layer(f"transformer_layer_{layer_idx}")
+        ._self_attention_layer._query_dense
+    ) = LoraLayer(
+        original_query_layer,
+        rank=RANK,
+        alpha=ALPHA,
+        layer_idx=layer_idx,
+        trainable=True,
+    )
+
+    # Change value dense layer.
+    original_value_layer = (
+        lora_model.get_layer(gpt2_backbone_layer_name)
+        .get_layer(f"transformer_layer_{layer_idx}")
+        ._self_attention_layer._value_dense
+    )
+
+    (
+        lora_model.get_layer(gpt2_backbone_layer_name)
+        .get_layer(f"transformer_layer_{layer_idx}")
+        ._self_attention_layer._value_dense
+    ) = LoraLayer(
+        original_value_layer,
+        rank=RANK,
+        alpha=ALPHA,
+        layer_idx=layer_idx,
+        trainable=True,
+    )
+
+"""
+Let's now do a forward pass to make sure it is going through.
```

Comment:
it is going through. -> we still have a valid chain of computation? or something like that